plugins/uefi-dbx/fu-uefi-dbx-snapd-notifier.c

```diff
@@ -0,0 +1,245 @@
+}
+
+static void
+fu_uefi_dbx_snapd_notifier_init(FuUefiDbxSnapdNotifier *self)
+{
+	/* default path for use inside the snap sandbox */
+	const char *snapd_snap_socket = "/run/snapd-snap.socket";
+	const char *snapd_snap_socket_override = g_getenv("FWUPD_SNAPD_SNAP_SOCKET");
+
+	self->curl_template = curl_easy_init();
+
+	/* TODO support system wide snapd socket for outside of snap scenarios */
+	if (!fu_snap_is_in_snap())
+		g_warning("attempted use of snapd notifier outside of snap");
+
+	if (snapd_snap_socket_override != NULL)
+		snapd_snap_socket = snapd_snap_socket_override;
+
+	/* use snap dedicated socket when running inside a snap */
+	curl_easy_setopt(self->curl_template, CURLOPT_UNIX_SOCKET_PATH, snapd_snap_socket);
+
+	self->req_hdrs = curl_slist_append(self->req_hdrs, "Content-Type: application/json");
+	curl_easy_setopt(self->curl_template, CURLOPT_HTTPHEADER, self->req_hdrs);
+}
+
+static void
+fu_uefi_dbx_snapd_notifier_finalize(GObject *object)
+{
+	FuUefiDbxSnapdNotifier *self = FU_UEFI_DBX_SNAPD_NOTIFIER(object);
+	curl_slist_free_all(self->req_hdrs);
+	curl_easy_cleanup(self->curl_template);
+	G_OBJECT_CLASS(fu_uefi_dbx_snapd_notifier_parent_class)->finalize(object);
+}
+
+static void
+fu_uefi_dbx_snapd_notifier_class_init(FuUefiDbxSnapdNotifierClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS(klass);
+	object_class->finalize = fu_uefi_dbx_snapd_notifier_finalize;
+}
+
+/* see CURLOPT_WRITEFUNCTION(3) */
+static size_t
+fu_uefi_dbx_snapd_notifier_rsp_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+	GByteArray *bufarr = (GByteArray *)userdata;
+	gsize sz = size * nmemb;
+	g_byte_array_append(bufarr, (const guint8 *)ptr, sz);
+	return sz;
+}
+
+static gboolean
+fu_uefi_dbx_snapd_notifier_simple_req(FuUefiDbxSnapdNotifier *self,
+				      const char *endpoint,
+				      const char *data,
+				      gsize len,
+				      GError **error)
+{
+	CURL *curl = NULL;
+	CURLcode res = -1;
+	glong status_code = 0;
+	g_autofree gchar *endpoint_str = NULL;
+	g_autoptr(GByteArray) rsp_buf = g_byte_array_new();
+
+	/* duplicate a preconfigured curl handle */
+	curl = curl_easy_duphandle(self->curl_template);
+
+	endpoint_str = g_strdup_printf("http://localhost%s", endpoint);
+
+	curl_easy_setopt(curl, CURLOPT_URL, endpoint_str);
+
+	curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, len);
+	curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+
+	/* collect response for debugging */
+	curl_easy_setopt(curl, CURLOPT_WRITEDATA, rsp_buf);
+	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fu_uefi_dbx_snapd_notifier_rsp_cb);
+
+	/* TODO: disable verbose or provide a handler */
+	curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+
+	res = curl_easy_perform(curl);
+	if (res != CURLE_OK) {
+		/* TODO inspect the error */
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "failed to communicate with snapd: %s",
+			    curl_easy_strerror(res));
+		return FALSE;
```

Comment:
I think you're leaking `curl` here -- it might be better to use G_DEFINE_AUTOPTR_CLEANUP_FUNC for the `CURL` like we have elsewhere in fwupd.